walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,99 @@
+    // Find availabilities for the mandatory attendees.
+    if (allAttendeesAvailabilities.size() == 0) {
+      return findAvailability(events, request, request.getAttendees());
+    }
+
+    return allAttendeesAvailabilities;
+  }
+
+  /** Private helper method to find all availabilities of people from the request. */
+  private Collection<TimeRange> findAvailability(
+      Collection<Event> events, MeetingRequest request, Collection<String> attendees) {
+    // Find all the unavailable times of the attendees.
+    List<TimeRange> unavailabilities = new ArrayList<>();
+    for (Event event : events) {
+      if (attendees.stream().anyMatch(s -> event.getAttendees().contains(s))) {
+        unavailabilities.add(event.getWhen());
+      }
+    }
+    // If there are no unavailable times, return whole day.
+    if (unavailabilities.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    List<TimeRange> mergedUnavailabilities =
+        mergeAllUnavailabilities((ArrayList<TimeRange>) unavailabilities);
+
+    // Find all available times by checking if gaps between busy time ranges are greater than the
+    // duration.
+    List<TimeRange> availabilities = new ArrayList<>();
+    int start = TimeRange.START_OF_DAY;
+    long duration = request.getDuration();
+    for (TimeRange interval : mergedUnavailabilities) {
+      if (interval.start() - start >= duration) {
+        availabilities.add(TimeRange.fromStartEnd(start, interval.start(), false));
+      }
+      start = interval.end();
+    }
+    // Check if there is a lot between there and the end of the day.
+    if (TimeRange.END_OF_DAY - start >= duration) {
+      availabilities.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true));
+    }
+    return availabilities;
+  }
+
+  /**
+   * Private helper method using Stack to merge overlapping time ranges from an ArrayList.
+   * Merging code from https://www.geeksforgeeks.org/merging-intervals/
+   */
+  private List<TimeRange> mergeAllUnavailabilities(ArrayList<TimeRange> unavailabilities) {
+    Stack<TimeRange> unavailabilitiesStack = new Stack<>();
+    unavailabilities.sort(TimeRange.ORDER_BY_START);
+
+    // Push the first unavaibility into the stack.
+    unavailabilitiesStack.push(unavailabilities.get(0));
+
+    // Start from the next unavailability and merge if necessary.
+    for (int i = 1; i < unavailabilities.size(); i++) {
+      TimeRange top = unavailabilitiesStack.peek();
+
+      // If current time range is not overlapping with the top, push it to the stack.
+      if (!top.overlaps(unavailabilities.get(i))) {
+        unavailabilitiesStack.push(unavailabilities.get(i));
+      } else if (top.contains(unavailabilities.get(i))) {
+        continue;
+      } else {
+        unavailabilitiesStack.pop();
+        unavailabilitiesStack.push(
+            TimeRange.fromStartEnd(top.start(), unavailabilities.get(i).end(), false));
```

Comment:
We call _unavailabilities.get(i)_ multiple times in this section. Perhaps it should be extracted to a variable (such as _currentMeeting_) to facilitate readability end emphasize you are inspecting only a single meeting in the loop.

Reply:
Done